File: training-service/src/models/annotation.py
from flask_mongoengine import MongoEngine

db = MongoEngine()


class AnnotationKey(db.EmbeddedDocument):
    language = db.StringField(required=True)
    lexingTokens = db.ListField(required=True)

class Annotation(db.Document):
    key = db.EmbeddedDocumentField(AnnotationKey, primary_key=True)
    sourceCode = db.StringField()
    highlightingTokens = db.ListField()
    highlightingCode = db.StringField()
    trainedTime = db.DateTimeField()
    validatedTime = db.DateTimeField()
    meta = {
        'collection': 'annotations',
        'strict': False
    }

File: annotation-service/src/main/java/com/hack3rz/annotationservice/model/Annotation.java
package com.hack3rz.annotationservice.model;

import com.hack3rz.annotationservice.enumeration.SupportedLanguage;
import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.List;

/**
 * Model for the annotations collection in the database
 */
@Document("annotations")
@Builder
@Getter
public class Annotation {

    @Id
    private AnnotationKey id;

    @Value
    @AllArgsConstructor
    public static class AnnotationKey {
        private SupportedLanguage language;
        private List<Integer> lexingTokens;
    }

    private String sourceCode;

    private List<Integer> highlightingTokens;

    private String highlightingCode;
}
